### PR TITLE
OCPBUGS-3009: Prune stale CCRs before aggregating scan results

### PR DESCRIPTION
**Pull request by rhmdnd**
Previously, the compliance operator would leave CCRs around and then
just overwrite them on subsequent scans. While the most recent scan data
was accurate, because it was overwriting existing check results, it gave
the impression that some changes weren't taking effect.

For example, if you create a tailored profile, run a scan, exclude a
rule, and rerun the scan, it appears the change you just made never took
effect because the result from the rule you ignored still exists.

To avoid this, let's check for any check results at scan time and make
sure we clean them up before we aggregate the new results.

### CODE REVIEW

**Comment by rhmdnd** [comment]
This still needs some testing but I'm waiting on a cluster.

File: config/manager/kustomization.yaml
Comment (Vincent056) [inline config/manager/kustomization.yaml:6]: could you remove this change from the commit?

**Comment by Vincent056** [comment]
I think it makes senses to remove CCR before aggreagating phase, but I wonder if we should do that during phaseDoneHandler, when someone labels a scan to be rescanned.

**Comment by rhmdnd** [comment]
I think this PR still needs some work, including an e2e test. I was in the process of adding a test and then discovered other issues in the test framework that were interfering with my test (specifically around resource cleanup).

**Comment by rhmdnd** [comment]
I was able to reproduce with the latest test.

**Comment by rhmdnd** [comment]
@xiaojiey I got a clean parallel run with the test. Should be good for another pre-merge validation test.

**Review by mkumku — COMMENTED** [review]
lgtm, thank you!

**Comment by rhmdnd** [comment]
/retest

More timeout issues.

**Comment by jhrozek** [comment]
Added qe-approved. This was tested.

**Review by jhrozek — APPROVED** [review]
/lgtm

**Comment by rhmdnd** [comment]
Rebased to resolve conflict in test utilities.

**Review by Vincent056 — APPROVED** [review]
/lgtm

**Comment by jhrozek** [comment]
/retest
timeout during test

**Comment by rhmdnd** [comment]
/test e2e-aws-serial

Error due to:

```
 2023/04/18 10:44:59 ComplianceScan ready (DONE)
2023/04/18 10:44:59 All scans in ComplianceSuite have finished (test-remediate)
2023/04/18 10:44:59 scan test-remediate-e2e re-run has finished
2023/04/18 10:44:59 Removing applied remediation
    main_test.go:445: complianceremediations.compliance.openshift.io "test-remediate-e2e-no-direct-root-logins" not found
--- FAIL: TestAutoRemediate (225.19s) 
```

Which doesn't seem related to this test but could be transient.

**Comment by rhmdnd** [comment]
This is probably a legitimate issue since the CCR is getting cleaned up, which removes the remediation. Looking into this to see if we need to update the tests or if this points to a bigger issue with the patch.

**Comment by rhmdnd** [comment]
/hold

The serial test failure is legitimate since remediations are owned by check results. When the scan reruns, it deletes the automatically applied remediation.

This raises at least two concerns.

1.) Do we really want to prune all compliance chekc results for a scan in the done phase?
2.) Should the lifecycle of a remediation be associated to the compliance check result?

**Comment by rhmdnd** [comment]
This was reported as an issue in https://github.com/ComplianceAsCode/compliance-operator/issues/73.

**Comment by rhmdnd** [comment]
@Vincent056 @yuumasato this should be ready for another review.

I optimized the logic so that we only clean up stale results, instead of deleting all the results when a scan is rescheduled. This prevents remediations from disappearing even though the check wasn't ignored.

**Comment by rhmdnd** [comment]
I wasn't able to reproduce the latest failures on a local cluster.

**Comment by rhmdnd** [comment]
@xiaojiey should be ready for pre-merge verification

File: tests/e2e/parallel/main_test.go
Comment (Vincent056) [inline tests/e2e/parallel/main_test.go:2921]: I wonder if we should check for a rule that generates the remediation, so we know that the remediation is also removed.

**Review by Vincent056 — APPROVED** [review]
/lgtm
just some inline comments but I think we can add that in another pr

File: tests/e2e/parallel/main_test.go
Comment (rhmdnd) [reply tests/e2e/parallel/main_test.go:2927]: Good call. Fixed and added a couple assertions to make sure the remediation is gone, too.

**Review by Vincent056 — APPROVED** [review]
/lgtm

**Comment by xiaojiey** [comment]
Pre-merge testing passed with 4.14.0-0.nightly-2023-11-09-092851 and code in https://github.com/ComplianceAsCode/compliance-operator/pull/221
```
1. Create a tp and create a ssb with the tailoredprofile:

$ cat tp2.yaml 
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: test
  namespace: openshift-compliance
spec:
  description: test
  title: test
  enableRules:
    - name: ocp4-api-server-anonymous-auth
      rationale: platform
    - name: ocp4-api-server-api-priority-gate-enabled
      rationale: platform
    - name: ocp4-api-server-audit-log-maxbackup
      rationale: platform
    - name: ocp4-api-server-audit-log-maxsize
      rationale: platform
    - name: ocp4-api-server-encryption-provider-cipher
      rationale: platform
$ oc apply -f tp2.yaml 
tailoredprofile.compliance.openshift.io/test created
]$ oc get tp
NAME   STATE
test   READY
$ oc compliance bind -N test tailoredprofile/test
Creating ScanSettingBinding test
$ oc compliance bind -N test tailoredprofile/test
Creating ScanSettingBinding test
$ oc get suite
NAME   PHASE     RESULT
test   DONE          NON-COMPLIANT
$ oc get ccr
NAME                                         STATUS   SEVERITY
test-api-server-anonymous-auth               PASS     medium
test-api-server-api-priority-gate-enabled    FAIL     medium
test-api-server-audit-log-maxbackup          PASS     low
test-api-server-audit-log-maxsize            PASS     medium
test-api-server-encryption-provider-cipher   FAIL     medium
$ oc get cr
NAME                                         STATE
test-api-server-encryption-provider-cipher   NotApplied
2. Edit to replace the rule ocp4-api-server-encryption-provider-cipher with ocp4-audit-profile-set, and rerun the ssb. Check the result for newly added rule is created, the removed rule has been remove from the ccr and cr.

$ oc get tp test -o yaml
...
spec:
  description: test
  enableRules:
  - name: ocp4-api-server-anonymous-auth
    rationale: platform
  - name: ocp4-api-server-api-priority-gate-enabled
    rationale: platform
  - name: ocp4-api-server-audit-log-maxbackup
    rationale: platform
  - name: ocp4-api-server-audit-log-maxsize
    rationale: platform
  - name: ocp4-audit-profile-set
    rationale: platform
  title: test
status:
  id: xccdf_compliance.openshift.io_profile_test
  outputRef:
    name: test-tp
    namespace: openshift-compliance
  state: READY
$ oc compliance rerun-now scansettingbinding test
Rerunning scans from 'test': test
Re-running scan 'openshift-compliance/test'
$ oc get suite
NAME   PHASE     RESULT
test   DONE          NON-COMPLIANT
$ oc get ccr
NAME                                        STATUS   SEVERITY
test-api-server-anonymous-auth              PASS     medium
test-api-server-api-priority-gate-enabled   FAIL     medium
test-api-server-audit-log-maxbackup         PASS     low
test-api-server-audit-log-maxsize           PASS     medium
test-audit-profile-set                      PASS     medium
$ oc get ccr --no-headers |wc -l
5
$ oc get cr
No resources found in openshift-compliance namespace.
##3. Add "extends: ocp4-cis" in the tp, and rerun the test result:

$ oc get tp test -o=jsonpath={.spec} | jq -r
{
  "description": "test",
  "enableRules": [
    {
      "name": "ocp4-api-server-anonymous-auth",
      "rationale": "platform"
    },
    {
      "name": "ocp4-api-server-api-priority-gate-enabled",
      "rationale": "platform"
    },
    {
      "name": "ocp4-api-server-audit-log-maxbackup",
      "rationale": "platform"
    },
    {
      "name": "ocp4-api-server-audit-log-maxsize",
      "rationale": "platform"
    },
    {
      "name": "ocp4-audit-profile-set",
      "rationale": "platform"
    }
  ],
  "extends": "ocp4-cis",
  "title": "test"
}
$ oc compliance rerun-now scansettingbinding test
Rerunning scans from 'test': test
Re-running scan 'openshift-compliance/test'
$ oc get suite -w
NAME   PHASE     RESULT
test   DONE          NON-COMPLIANT
$ oc get ccr --no-headers |wc -l
88
$ oc get ccr | head
NAME                                                          STATUS   SEVERITY
test-accounts-restrict-service-account-tokens                 MANUAL   medium
test-accounts-unique-service-account                          MANUAL   medium
test-api-server-admission-control-plugin-alwaysadmit          PASS     medium
test-api-server-admission-control-plugin-alwayspullimages     PASS     high
test-api-server-admission-control-plugin-namespacelifecycle   PASS     medium
test-api-server-admission-control-plugin-noderestriction      PASS     medium
test-api-server-admission-control-plugin-scc                  PASS     medium
test-api-server-admission-control-plugin-service-account      PASS     medium
test-api-server-anonymous-auth                                PASS     medium
```

**Comment by Vincent056** [comment]
I was trying to resolve the merge conflicts through github UI and then it got the master to merge to the branch to commit here

**Comment by rhmdnd** [comment]
Removed the hold flag since we have QE approval.

I also rebased locally to resolve the merge conflict with master in the same patch.

Thanks for all the reviews folks.

**Review by yuumasato — APPROVED** [review]
/lgtm

**Pull request merged**
